### PR TITLE
Swig bindings for RigidBodyTree

### PR DESCRIPTION
**Pull request by rdeits**
This is a more focused version of #1544 which just adds some initial bindings for RigidBodyTree and IK. Currently, everything works in both Python and Matlab, which is really exciting. The packaging is (by necessity) slightly different in the two languages: the Python bindings are all organized into a `pydrake` hierarchy that can be installed like a normal Python package, while the Matlab bindings are in a set of smaller packages `ik`, `rbtree`, etc which get built into the `pod-build/matlab` folder. 

I have some very very limited unit tests in place to verify the functionality. See `drake/bindings/python/pydrake/test/` and `drake/bindings/matlab` for the tests and to compare the two APIs. 

This is all likely to be subject to change in the near future, but this is the first time I feel like we have something we can actually use in both languages.

### CODE REVIEW

**Comment by rdeits** [comment]
Note that there's currently no way this will work on Windows, since building Swig requires a unix-type environment (and we don't require cygwin or equivalent). This seems like something that our new collaborators could potentially help with...

**Comment by rdeits** [comment]
@RussTedrake I can't tell what's going on with drake005's python tests. I can run the same command that (I think) ctest is running, and it passes, but it looks like it might just be failing to decode the test output? I don't actually know what `decode.py` does. 

The good news is that the new python bindings and their unit tests are passing on drake006 and all four Travis builds, and I've successfully dodged the issue on Windows for now. 

It looks like the Matlab unit tests aren't being picked up yet, though, so I've got a little more work to do.

**Comment by rdeits** [comment]
@RussTedrake can you also please run the `install_prereqs` script on drake001 or just `brew install autoconf automake libtool`?

**Comment by RussTedrake** [comment]
decode.py unpacks the compressed test output stored by ctest.  it is a failure of the unit test system, not a failure of your code.  i will add a known issue about it.  
i've run the brew install line on drake001.

**Comment by rdeits** [comment]
Great, thanks! Hopefully this time will be all clean

**Comment by RussTedrake** [comment]
actually.. if you have reliable decode.py failures, it could be due to something like a strange character in the console output from your test?

**Comment by rdeits** [comment]
Ah, it looks like a real issue (due to the fact that cmake inexplicably really wants to use /usr/bin/python instead of /usr/local/bin/python on drake005. But the error message doesn't have any special characters: 

```
Fatal Python error: PyThreadState_Get: no current thread
```

Anyway, I'm attempting to fix the underlying error now.

**Comment by rdeits** [comment]
@RussTedrake this is very odd. It looks like on drake001 and drake005, cmake is always choosing the system-provided python from /usr/bin instead of /usr/local/bin (the homebrew version). That's a problem because the homebrew python is the one that we actually want (it has the development headers, the installed packages, etc). This is also the root cause of the problems I was having with the BeautifulSoup dependency in the textbook examples PR. When I ssh into drake005, su to your user, and run `make` myself, cmake appears to detect the correct version. In addition, the Travis mac build servers (which use homebrew) work just fine, and my personal machine works fine. So it's something odd about the way the drake unit testing framework invokes make. 

I've created a workaround at the top level to force the PYTHON_EXECUTABLE to be the one that's found on the path on OSX: https://github.com/RobotLocomotion/drake/commit/8183d6363cf606204f2fc55dfa88b5b9aaff24d2 but this may be a problem in the future too.

**Comment by rdeits** [comment]
Aha! It looks like the drake unit testing framework actually does not have `/usr/local/bin` in its PATH (so my workaround in 8183d6363cf606204f2fc55dfa88b5b9aaff24d2 was both unnecessary and unhelpful. That's definitely going to be a problem, but should be easy to fix. Is your unit test script perhaps not sourcing your personal `.bash_profile`?

File: CMakeLists.txt [hunk 1]
```diff
@@ -69,6 +74,10 @@ if (NOT WIN32) # many of these might work on win32 with little or no work... the
 	option(WITH_MESHCONVERTERS "uses vcglib to convert a few standard filetypes")
 endif()
 
+# The Drake mac build servers don't appear to have /usr/local/bin on their path, which causes lots of python problems. 
+if (APPLE)
+	set(ENV{PATH} "/usr/local/bin:$ENV{PATH}")
```
Comment (tkoolen) [inline CMakeLists.txt]: tab

File: drake/.UNITTEST [hunk 2]
```diff
@@ -1,4 +1,5 @@
-solvers	     TEST_DIRS_ONLY
-systems      TEST_DIRS_ONLY
-util         TEST_DIRS_ONLY
+solvers	        TEST_DIRS_ONLY
```
Comment (tkoolen) [inline drake/.UNITTEST]: tab

File: drake/bindings/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,17 @@
+if (NOT WIN32)
```
Comment (tkoolen) [inline drake/bindings/CMakeLists.txt]: Fix indentation throughout.

File: drake/bindings/python/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,11 @@
+function(add_python_test module_name)
```
Comment (tkoolen) [inline drake/bindings/python/CMakeLists.txt]: fix indentation.

File: drake/bindings/python/pydrake/test/testRBTCoM.py [hunk 5]
```diff
@@ -0,0 +1,19 @@
+from __future__ import print_function
```
Comment (tkoolen) [inline drake/bindings/python/pydrake/test/testRBTCoM.py]: So we're doing python 2? I hear that people are finally switching to python 3; should we just do python 3 from the start?

File: drake/bindings/swig/autodiffutils.i [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+%include <eigen.i>
+%include <autodiff.i>
+
+#ifdef SWIGPYTHON3
+%rename(__truediv__)     *::operator/;
+#endif
+
+%template(AutoDiffVectorDynamic) AutoDiffWrapper<Eigen::VectorXd, Eigen::Dynamic, 1>;
+%template(AutoDiffVectorMax73) AutoDiffWrapper<Eigen::Matrix<double, Eigen::Dynamic, 1, 0, 73>, Eigen::Dynamic, 1>;
+%template(AutoDiffMatrixDynamic) AutoDiffWrapper<Eigen::VectorXd, Eigen::Dynamic, Eigen::Dynamic>;
```
Comment (tkoolen) [inline drake/bindings/swig/autodiffutils.i]: How do these work?

File: drake/bindings/swig/autodiffutils.i [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+            derivatives = derivatives.reshape((-1,1))
+    if value.ndim == 0:
+        value = value.reshape((1,))
+    if value.ndim < 2 or value.shape[1] == 1:
+        # if derivatives.shape[1] <= 73:
+        #     return AutoDiffVectorMax73(value, derivatives)
+        # else:
+        return AutoDiffVectorDynamic(value, derivatives)
+    else:
+        return AutoDiffMatrixDynamic(value, derivatives)
```
Comment (tkoolen) [inline drake/bindings/swig/autodiffutils.i]: Why is there no AutoDiffMatrixMax73?

File: drake/bindings/swig/rbtree.i [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+
+%include <typemaps.i>
+%include <std_vector.i>
+%include <eigen.i>
+%include <autodiff.i>
+%import <autodiffutils.i>
+
+%template(vectorVectorXd) std::vector<Eigen::VectorXd>;
+%template(vectorMatrixXd) std::vector<Eigen::MatrixXd>;
+%template(vectorString) std::vector<std::string>;
```
Comment (tkoolen) [inline drake/bindings/swig/rbtree.i]: What do these do? I see no other references to `vectorVectorXd` etc.

File: drake/bindings/swig/rbtree.i [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+
+%eigen_typemaps(Eigen::VectorXd)
+%eigen_typemaps(Eigen::Matrix<double, SPACE_DIMENSION, 1>)
+%eigen_typemaps(Eigen::MatrixXd)
+%eigen_typemaps(Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic>)
+%eigen_typemaps(Eigen::VectorXi)
+
+%include "KinematicsCache.h"
+%template(KinematicsCache_d) KinematicsCache<double>;
+%template(KinematicsCache_adVectorDynamic) KinematicsCache<Eigen::AutoDiffScalar<Eigen::VectorXd> >;
```
Comment (tkoolen) [inline drake/bindings/swig/rbtree.i]: So no fixed max size yet?

File: drake/systems/plants/IKoptions.h [hunk 10]
```diff
@@ -51,7 +51,7 @@ class DRAKEIKOPTIONS_EXPORT IKoptions
     void setqd0(const Eigen::VectorXd &lb, const Eigen::VectorXd &ub);
     void setqdf(const Eigen::VectorXd &lb, const Eigen::VectorXd &ub);
     void setAdditionaltSamples(const Eigen::RowVectorXd &t_samples);
-    void updateRobot(const RigidBodyTree * robot);
+    void updateRobot(RigidBodyTree * new_robot);
```
Comment (tkoolen) [inline drake/systems/plants/IKoptions.h]: Why was this necessary?

File: drake/systems/plants/KinematicsCache.h [hunk 11]
```diff
@@ -47,7 +47,9 @@ class KinematicsCacheElement
   }
 
 public:
+#ifndef SWIG
 	EIGEN_MAKE_ALIGNED_OPERATOR_NEW
```
Comment (tkoolen) [inline drake/systems/plants/KinematicsCache.h]: tab

File: drake/systems/plants/KinematicsCache.h [hunk 12]
```diff
@@ -230,7 +232,9 @@ class KinematicsCache
...
```
Comment (tkoolen) [inline drake/systems/plants/KinematicsCache.h]: tab

File: drake/systems/plants/RigidBodyIK.h [hunk 13]
```diff
@@ -9,6 +9,18 @@
 
 class RigidBodyTree;
 
+class IKResults {
+public:
+  Eigen::VectorXd q_sol;
+  int INFO;
+  std::vector<std::string> infeasible_constraint;
```
Comment (tkoolen) [inline drake/systems/plants/RigidBodyIK.h]: infeasible_constraints?

File: drake/systems/plants/RigidBodyTree.h [hunk 14]
```diff
@@ -47,7 +47,9 @@ class DRAKERBM_EXPORT RigidBodyLoop
   friend std::ostream& operator<<(std::ostream& os, const RigidBodyLoop& obj);
 
 public:
+#ifndef SWIG 
 	EIGEN_MAKE_ALIGNED_OPERATOR_NEW
```
Comment (tkoolen) [inline drake/systems/plants/RigidBodyTree.h]: tab

File: drake/systems/plants/inverseKin.cpp [hunk 15]
```diff
@@ -17,3 +17,24 @@ DRAKEIK_EXPORT void inverseKin(RigidBodyTree * model, const MatrixBase<DerivedA>
...
 
+IKResults inverseKinSimple(RigidBodyTree* model, 
+  const Eigen::VectorXd &q_seed,
+  const Eigen::VectorXd &q_nom,
+  const std::vector<RigidBodyConstraint*> &constraint_array,
+  const IKoptions &ikoptions) {
+  auto results = IKResults();
+  results.q_sol.resize(q_nom.size());
+  int num_constraints = constraint_array.size();
+  RigidBodyConstraint** const constraint_array_ptr = (RigidBodyConstraint** const) &constraint_array[0];
```
Comment (tkoolen) [inline drake/systems/plants/inverseKin.cpp]: use .data().

**Comment by tkoolen** [comment]
Pretty cool!

File: CMakeLists.txt (hunk 1 above)
Comment (rdeits) [inline CMakeLists.txt]: that file is all tab-indented already

File: drake/bindings/python/pydrake/test/testRBTCoM.py (hunk 5 above)
Comment (rdeits) [inline drake/bindings/python/pydrake/test/testRBTCoM.py]: this code allows us to be agnostic (the same code will run correctly in python2 and python3. I've also verified that the basic capabilities work in python3, but stopped short of making it the default. I'm open to requiring python3, though. It might also solve the ridiculous path issues I'm having on the mac build servers.

File: drake/bindings/swig/rbtree.i (hunk 8 above)
Comment (rdeits) [inline drake/bindings/swig/rbtree.i]: We have to name them something for Swig to be happy, but it will correctly handle the conversion to python lists internally even without us using the name

File: drake/bindings/swig/autodiffutils.i (hunk 7 above)
Comment (tkoolen) [inline drake/bindings/swig/autodiffutils.i]: Ah so nothing really does fixed max size yet, correct?

File: drake/bindings/swig/rbtree.i (hunk 9 above)
Comment (rdeits) [inline drake/bindings/swig/rbtree.i]: No, but it's easy to add. I can do that before we merge if necessary

File: drake/systems/plants/IKoptions.h (hunk 10 above)
Comment (rdeits) [inline drake/systems/plants/IKoptions.h]: The header declared the method twice, once with const and once without. But the implementation does not have const, and Swig was getting confused by the duplicate (and incorrect) declaration.

File: drake/systems/plants/RigidBodyIK.h (hunk 13 above)
Comment (rdeits) [inline drake/systems/plants/RigidBodyIK.h]: should be, but it's singular everywhere else in the code

File: drake/bindings/swig/autodiffutils.i (hunk 6 above)
Comment (rdeits) [inline drake/bindings/swig/autodiffutils.i]: Implementation is here: https://github.com/rdeits/swigmake/blob/master/swig/common/autodiff.i

File: drake/bindings/swig/autodiffutils.i (hunk 7 above)
Comment (rdeits) [inline drake/bindings/swig/autodiffutils.i]: Just haven't turned it on yet. I can do that before we merge if necessary

File: drake/systems/plants/inverseKin.cpp (hunk 15 above)
Comment (rdeits) [inline drake/systems/plants/inverseKin.cpp]: oops, yeah, good point

File: drake/systems/plants/RigidBodyTree.h (hunk 14 above)
Comment (rdeits) [inline drake/systems/plants/RigidBodyTree.h]: fixed

File: drake/systems/plants/KinematicsCache.h (hunk 12 above)
Comment (rdeits) [inline drake/systems/plants/KinematicsCache.h]: fixed

File: drake/systems/plants/KinematicsCache.h (hunk 11 above)
Comment (rdeits) [inline drake/systems/plants/KinematicsCache.h]: fixed

File: drake/bindings/python/CMakeLists.txt (hunk 4 above)
Comment (rdeits) [inline drake/bindings/python/CMakeLists.txt]: fixed

File: drake/.UNITTEST (hunk 2 above)
Comment (rdeits) [inline drake/.UNITTEST]: fixed

File: drake/bindings/swig/autodiffutils.i (hunk 7 above)
Comment (rdeits) [inline drake/bindings/swig/autodiffutils.i]: That's right. The capability is there, but I'm trying to keep things simple for the first pass.

**Comment by rdeits** [comment]
Sigh. Yeah, it's definitely the path issue. When SSHed into drake005 after `su russt`, I see:

```
drake005% echo $PATH                                                                                                                    
/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/opt/X11/bin
```

which is correct. Within cmake, however, I see:

```
PATH: /usr/bin:/bin:/usr/sbin:/sbin:/usr/local/bin:/opt/X11/bin
```

@RussTedrake I don't think I know enough about what's going on inside the build servers to fix this. Can you take a look? We just need to make sure that /usr/local/bin is first.

**Comment by rdeits** [comment]
@tkoolen Okay, I've turned on fixed max-size 73 autodiffs for the python and matlab wrappers :-)

**Comment by RussTedrake** [comment]
Should be fixed now. (the path is specified manually in the crontab).  - Russ

**Comment by rdeits** [comment]
Thanks!

**Comment by rdeits** [comment]
@RussTedrake can you apply the fix to drake001, too?

**Comment by RussTedrake** [comment]
Done.  and manually restarted your test on drake001 specifically.

**Comment by rdeits** [comment]
Okay, thanks again. I've cleaned up a few things and removed some unnecessary changes.

**Comment by rdeits** [comment]
Oh, come on... It looks like Gurobi has changed their mac python installer, but I'm only getting the new version some of the time (and the new version doesn't work). Compare the `setup.py` that the build server downloaded: 

```
if sys.platform == 'darwin':                                                                                   
  import platform,subprocess                                                                                   
  from distutils.sysconfig import get_python_lib                                                               
  sitelib = get_python_lib() + '/gurobipy/gurobipy.so'                                                         
  if platform.mac_ver()[0] >= "10.11": # El Capitan                                                            
    import glob,distutils.spawn                                                                                
    GUROBI_HOME = os.sep.join(os.readlink(distutils.spawn.find_executable("gurobi_cl")).split(os.sep)[:-2])    
    SHARED_LIB = glob.glob(GUROBI_HOME+"/lib/libgurobi*.so")[0].split(os.sep)[-1]                              
    subprocess.call(('install_name_tool', '-change',                                                           
                     SHARED_LIB, GUROBI_HOME + "/lib/" + SHARED_LIB, sitelib))                                 
  default = '/System/Library/Frameworks/Python.framework/Versions/2.7/Python'                                  
  modified = sys.prefix + '/Python'                                                                            
  if default != modified:                                                                                      
    import os.path                                                                                             
    if not os.path.isfile(modified):                                                                           
      modified = sys.prefix + '/lib/libpython2.7.dylib' # For Anaconda                                         
    subprocess.call(('install_name_tool', '-change', default, modified, sitelib)) 
```

and the `setup.py` that I got by running a new build on the same machine: 

```
if sys.platform == 'darwin':                                                                                   
  default = '/System/Library/Frameworks/Python.framework/Versions/2.7/Python'                                  
  modified = sys.prefix + '/Python'                                                                            
  if default != modified:                                                                                      
    from distutils.sysconfig import get_python_lib                                                             
    import subprocess                                                                                          
    import os.path                                                                                             
    if not os.path.isfile(modified):                                                                           
      modified = sys.prefix + '/lib/libpython2.7.dylib' # For Anaconda                                         
    sitelib = get_python_lib() + '/gurobipy/gurobipy.so'                                                       
    subprocess.call(('install_name_tool', '-change', default, modified, sitelib)) 
```

That's why drake001 is currently failing: it got the new gurobi installer instead of the old one. I'll try to fix it.

**Comment by rdeits** [comment]
Also, the new version is not just different, but totally wrong. It tries to check for OSX 10.11 by doing:

```
platform.mac_ver()[0] >= "10.11" 
```

but comparison operators on strings are *definitely* not the right thing to do, so that check actually evals to `True` on drake001 (even though it's on 10.9.4):

```
>>> platform.mac_ver()[0]                                                                                      
'10.9.4'                                                                                                       
>>> platform.mac_ver()[0] >= "10.11"                                                                           
True                                                                                                               
```

Derp.

**Comment by rdeits** [comment]
Closing and re-opening fixed the gurobi build issue, so I think we're good to go once drake004 reports

File: drake/bindings/matlab/+autodiffutils/newAutoDiff.m [hunk 16]
```diff
@@ -0,0 +1,18 @@
+function obj = newAutoDiff(value, derivatives)
```
Comment (RussTedrake) [inline drake/bindings/matlab/+autodiffutils/newAutoDiff.m]: naming things "new" is a bad idea -- no context and will soon be out of date.

File: drake/bindings/python/pydrake/test/testRBTCoM.py (hunk 5 above)
Comment (RussTedrake) [inline drake/bindings/python/pydrake/test/testRBTCoM.py]: just for the record -- pretty sure we've resolved those path issues (it wasn't due to the python 2 vs 3, but the path)

File: drake/systems/plants/KinematicsCache.h [hunk 17]
```diff
@@ -47,7 +47,9 @@ class KinematicsCacheElement
   }
 
 public:
-	EIGEN_MAKE_ALIGNED_OPERATOR_NEW
+#ifndef SWIG
+  EIGEN_MAKE_ALIGNED_OPERATOR_NEW
```
Comment (RussTedrake) [inline drake/systems/plants/KinematicsCache.h]: does this mean that it will not work on a 32-bit OS?

File: drake/systems/plants/constraint/RigidBodyConstraint.h [hunk 18]
```diff
@@ -427,7 +432,9 @@ class DRAKERIGIDBODYCONSTRAINT_EXPORT GazeConstraint : public SingleTimeKinemati
     virtual ~GazeConstraint(void){};
 
 public:
+#ifndef SWIG
```
Comment (RussTedrake) [inline drake/systems/plants/constraint/RigidBodyConstraint.h]: if this is going to appear everywhere, and it's just wrapping a macro, why not make a new macro that includes the swig logic.  (in fact eigen does have EIGEN_MAKE_ALIGNED_OPERATOR_NEW_IF(NeedsToAlign) already defined)

**Comment by RussTedrake** [comment]
hoorah!  clean builds.  I've done my code review and have just a few small comments.

File: drake/bindings/matlab/+autodiffutils/newAutoDiff.m (hunk 16 above)
Comment (rdeits) [inline drake/bindings/matlab/+autodiffutils/newAutoDiff.m]: it's actually "new" because it invokes a constructor, but I see your point. I'll rename it to something clearer.

File: drake/systems/plants/KinematicsCache.h (hunk 17 above)
Comment (rdeits) [inline drake/systems/plants/KinematicsCache.h]: No (although I was originally concerned about that too). The SWIG symbol is just defined for the Swig preprocessor which scans the header files to figure out what to wrap. The C++ interface code that Swig creates still links against the original Drake libraries with the correct eigen allocators. For some reason, Eigen's macros were just confusing the preprocessor.

File: drake/systems/plants/constraint/RigidBodyConstraint.h (hunk 18 above)
Comment (rdeits) [inline drake/systems/plants/constraint/RigidBodyConstraint.h]: ok, can do

File: drake/bindings/matlab/+autodiffutils/newAutoDiff.m (hunk 16 above)
Comment (rdeits) [inline drake/bindings/matlab/+autodiffutils/newAutoDiff.m]: changed to `toAutoDiff`

File: drake/systems/plants/constraint/RigidBodyConstraint.h (hunk 18 above)
Comment (rdeits) [inline drake/systems/plants/constraint/RigidBodyConstraint.h]: Actually, this turns out to be harder than you'd think, because in C++ you can't create a macro which contains an `#ifdef` directive. I think unless this is something we want to spend significant time on, we might just have to live with it.

File: drake/bindings/matlab/+autodiffutils/toAutoDiff.m [hunk 19]
```diff
@@ -0,0 +1,18 @@
+function obj = toAutoDiff(value, derivatives)
+
+if nargin < 2
+  derivatives = eye(numel(value));
+end
+
+if size(value, 2) == 1
+  if size(value, 1) <= 73
+    obj = autodiffutils.AutoDiffVectorMax73(value, derivatives);
```
Comment (RussTedrake) [inline drake/bindings/matlab/+autodiffutils/toAutoDiff.m]: am i missing something?  how does this work?  I don't see that method in the autodiffutils folder?
also, is bindings/matlab definitely the right place for this?  it's too bad to add the directory to the path just for this one file.  perhaps it should live under util?

Comment (rdeits) [inline drake/bindings/matlab/+autodiffutils/toAutoDiff.m]: The matlab files that Swig generates get put in a `+autodiffutils` package inside pod-build, which is added to the Matlab path. It turns out that you can have multiple folders with the same package name in Matlab, and all of them will get searched for the appropriate methods, so this file looks to the users like its in the same package as the files it references.
